src/core/rate_limiter.h

```diff
@@ -0,0 +1,263 @@
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <condition_variable>
+#include <functional>
+#include <mutex>
+#include <queue>
+#include <vector>
+
+#include "src/core/model_config.pb.h"
+#include "src/core/status.h"
+
+namespace nvidia { namespace inferenceserver {
+
+// Limits the rate at which requests are dispatched from the scheduler
+class RateLimiter {
+ private:
+  class ModelContext;
+
+ public:
+  class ModelInstance;
+
+  using StandardReleaseFunc = std::function<void(ModelInstance*)>;
+  using StandardScheduleFunc = std::function<void(ModelInstance*)>;
+  using StandardStageFunc = std::function<void(ModelInstance*)>;
+  using RateLimiterConfig = inference::ModelRateLimiter;
+
+  /// Creates a rate limiter object which will funnel the requests to
+  /// the model instances. A typical lifetime of the model instance within
+  /// RateLimiter transition from available -> staged -> allocated -> available.
+  /// The transition from available to staged occurs when a request is
+  /// registered for the model. Depending upon the resource availabilty and
+  /// priority, the RateLimiter will transition an instance to allocated state
+  /// at some point in the future. The staged state is skipped when
+  /// configured to ignore the resource constraints. The cycle in this case
+  /// will be available -> allocated -> available.
+  /// \param ignore_resources_and_priority Whether or not to ignore resource
+  /// constraints and cross-model priority. An available instance is directly
+  /// allocated when true.
+  /// \return Status object indicating success or failure.
+  static Status Create(
+      const bool ignore_resources_and_priority,
+      std::unique_ptr<RateLimiter>* rate_limiter);
+
+  /// Loads the configified model to the RateLimiter
+  /// \param model_name The name of the model to load.
+  /// \param version The version of the model to load.
+  /// \param model_config The configuration of the model to load.
+  /// \return Status object indicating success or failure.
+  Status LoadModel(
+      const std::string& model_name, const int64_t version,
+      const inference::ModelConfig& model_config);
+
+  /// Unloads the configified model from the RateLimiter
+  /// \param model_name The name of the model to load.
+  /// \param version The version of the model to load.
+  /// \return Status object indicating success or failure.
+  Status UnloadModel(const std::string& model_name, const int64_t version);
+
+  /// Enqueues the callback to the configified model. In future, when the
```

Comment:
Not sure what "configified" means or if it is important. And isn't what is being requested is a model instance? So perhaps "RequestModelInstance"?